### PR TITLE
Cable net/new fixes

### PR DESCRIPTION
**Pull request by KlausBSautter**
- as discussed with @philbucher I moved a test which took to long
- added the possibility to use the sliding cable element with plasticity

### CODE REVIEW

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp [hunk 1]
```diff
@@ -92,9 +92,15 @@ void SlidingCableElement3D::GetDofList(DofsVectorType &rElementalDofList,
...
 }
 
-void SlidingCableElement3D::Initialize() {
-  KRATOS_TRY
-  KRATOS_CATCH("")
+void SlidingCableElement3D::Initialize()
+{
+    KRATOS_TRY
+    if (GetProperties()[CONSTITUTIVE_LAW] != nullptr) {
+        mpConstitutiveLaw = GetProperties()[CONSTITUTIVE_LAW]->Clone();
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]: don't save it extra, use `Check` to check if it exists and the retrieve it from the Properties when you need it

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp [hunk 2]
```diff
@@ -425,13 +445,17 @@ Vector SlidingCableElement3D::GetCustomInternalForceWithFriction(const Vector& r
 }
 
 
-Matrix SlidingCableElement3D::ElasticStiffnessMatrix()
+Matrix SlidingCableElement3D::ElasticStiffnessMatrix(ProcessInfo& rCurrentProcessInfo)
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]:
```suggestion
Matrix SlidingCableElement3D::ElasticStiffnessMatrix(const ProcessInfo& rCurrentProcessInfo)
```

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp [hunk 3]
```diff
@@ -442,7 +466,7 @@ Matrix SlidingCableElement3D::ElasticStiffnessMatrix()
   return elastic_stiffness_matrix;
 }
 
-Matrix SlidingCableElement3D::GeometricStiffnessMatrix()
+Matrix SlidingCableElement3D::GeometricStiffnessMatrix(ProcessInfo& rCurrentProcessInfo)
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]:
```suggestion
Matrix SlidingCableElement3D::GeometricStiffnessMatrix(const ProcessInfo& rCurrentProcessInfo)
```

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp [hunk 4]
```diff
@@ -494,10 +524,10 @@ Matrix SlidingCableElement3D::GeometricStiffnessMatrix()
   return geometric_stiffness_matrix;
 }
 
-inline Matrix SlidingCableElement3D::TotalStiffnessMatrix()
+inline Matrix SlidingCableElement3D::TotalStiffnessMatrix(ProcessInfo& rCurrentProcessInfo)
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]:
```suggestion
inline Matrix SlidingCableElement3D::TotalStiffnessMatrix(const ProcessInfo& rCurrentProcessInfo)
```

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp [hunk 5]
```diff
@@ -749,6 +779,10 @@ int SlidingCableElement3D::Check(const ProcessInfo& rCurrentProcessInfo)
     const double domain_size = this->GetCurrentLength();
     KRATOS_ERROR_IF( domain_size <= 0.0 ) << "Element " << this->Id() << " has non-positive size " << domain_size << std::endl;
 
+    if (mpConstitutiveLaw != nullptr) {
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]: remove the check here, it is enough to check it once in `Check`

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp [hunk 6]
```diff
@@ -844,10 +878,59 @@ Vector SlidingCableElement3D::CalculateProjectionLengths()
...
+
+
+
 void SlidingCableElement3D::save(Serializer &rSerializer) const {
   KRATOS_SERIALIZE_SAVE_BASE_CLASS(rSerializer, Element);
+    rSerializer.save("mpConstitutiveLaw", mpConstitutiveLaw);
 }
 void SlidingCableElement3D::load(Serializer &rSerializer) {
   KRATOS_SERIALIZE_LOAD_BASE_CLASS(rSerializer, Element);
+  rSerializer.load("mpConstitutiveLaw", mpConstitutiveLaw);
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp:930]:
```suggestion

```

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp [hunk 7]
```diff
@@ -844,10 +878,59 @@ Vector SlidingCableElement3D::CalculateProjectionLengths()
...
+            GetGeometry(),temp_vector,true,true,rSaveInternalVariables);
+
+    KRATOS_CATCH("");
+}
+
+
+
 void SlidingCableElement3D::save(Serializer &rSerializer) const {
   KRATOS_SERIALIZE_SAVE_BASE_CLASS(rSerializer, Element);
+    rSerializer.save("mpConstitutiveLaw", mpConstitutiveLaw);
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp:926]:
```suggestion

```

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.hpp [hunk 8]
```diff
@@ -40,6 +40,8 @@ namespace Kratos
     {
     protected:
 
+        ConstitutiveLaw::Pointer mpConstitutiveLaw = nullptr;
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.hpp:43]:
```suggestion

```

**Review by philbucher — CHANGES_REQUESTED** [review]

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp (hunk 1 above)
Comment (tteschemacher) [reply applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]: If you want to have it historic on each gp you should clone it. I don't know if that is the case...

Comment (philbucher) [reply applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]: you are right, my bad, never mind then
Just move the check to `Check`

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp (hunk 5 above)
Comment (KlausBSautter) [reply applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]: ?

Comment (KlausBSautter) [reply applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]: I want to call the constitutive check function. Should I do this without 
`if (mpConstitutiveLaw != nullptr)`

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp [hunk 9]
```diff
@@ -749,6 +775,12 @@ int SlidingCableElement3D::Check(const ProcessInfo& rCurrentProcessInfo)
     const double domain_size = this->GetCurrentLength();
     KRATOS_ERROR_IF( domain_size <= 0.0 ) << "Element " << this->Id() << " has non-positive size " << domain_size << std::endl;
 
+    if (mpConstitutiveLaw != nullptr) {
+        mpConstitutiveLaw->Check(GetProperties(),GetGeometry(),rCurrentProcessInfo);
+    }
+
+    KRATOS_ERROR_IF(GetProperties()[CONSTITUTIVE_LAW] == nullptr)  << "A constitutive law needs to be specified for the element with ID " << Id() << std::endl;
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]:
```suggestion
    KRATOS_ERROR_IF_NOT(GetProperties()[CONSTITUTIVE_LAW])  << "A constitutive law needs to be specified for the element with ID " << Id() << std::endl;
```

I see that this is less readable, up to you

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp (hunk 5 above)
Comment (philbucher) [reply applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]: yes, move this check after the check for existance of the CLaw then you don't have to check twice

**Review by philbucher — COMMENTED** [review]
are you using the new interface of the CLaws?

**Comment by KlausBSautter** [comment]
> are you using the new interface of the CLaws?

I basically just copied the respective lines from the truss element

**Comment by philbucher** [comment]
could you please try the interface that is not deprecated?
This we should also change in structural at some point

**Comment by KlausBSautter** [comment]
> could you please try the interface that is not deprecated?
> This we should also change in structural at some point

I think this should do the job. I also changed the truss elements in the StructMechApp

**Comment by KlausBSautter** [comment]
`FinalizeNonLinearIteration` is not changed as I don't understand why this function is deprecated

**Comment by KlausBSautter** [comment]
@philbucher it is compiling now

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp [hunk 10]
```diff
@@ -425,13 +441,17 @@ Vector SlidingCableElement3D::GetCustomInternalForceWithFriction(const Vector& r
 }
 
 
-Matrix SlidingCableElement3D::ElasticStiffnessMatrix()
+Matrix SlidingCableElement3D::ElasticStiffnessMatrix(const ProcessInfo& rCurrentProcessInfo)
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]:
```suggestion
Matrix SlidingCableElement3D::ElasticStiffnessMatrix(const ProcessInfo& rCurrentProcessInfo) const
```

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp [hunk 11]
```diff
@@ -442,7 +462,7 @@ Matrix SlidingCableElement3D::ElasticStiffnessMatrix()
   return elastic_stiffness_matrix;
 }
 
-Matrix SlidingCableElement3D::GeometricStiffnessMatrix()
+Matrix SlidingCableElement3D::GeometricStiffnessMatrix(const ProcessInfo& rCurrentProcessInfo)
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]:
```suggestion
Matrix SlidingCableElement3D::GeometricStiffnessMatrix(const ProcessInfo& rCurrentProcessInfo) const
```

File: applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp [hunk 12]
```diff
@@ -494,10 +520,10 @@ Matrix SlidingCableElement3D::GeometricStiffnessMatrix()
   return geometric_stiffness_matrix;
 }
 
-inline Matrix SlidingCableElement3D::TotalStiffnessMatrix()
+inline Matrix SlidingCableElement3D::TotalStiffnessMatrix(const ProcessInfo& rCurrentProcessInfo)
```
Comment (philbucher) [inline applications/CableNetApplication/custom_elements/sliding_cable_element_3D.cpp]:
```suggestion
inline Matrix SlidingCableElement3D::TotalStiffnessMatrix(const ProcessInfo& rCurrentProcessInfo) const
```

**Review by philbucher — COMMENTED** [review]
please make the functions const if they don't modify class members

**Review by philbucher — APPROVED** [review]

**Pull request merged**